src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        static partial void TryCloneFile(string sourceFullPath, string destFullPath, bool overwrite, ref bool cloned)
+        {
+            // This helper function calls out to clonefile, and returns the error.
+            static bool TryCloneFile(string sourceFullPath, string destFullPath, int flags, out Interop.Error error)
+            {
+                if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+                {
+                    // Success.
+                    error = Interop.Error.SUCCESS;
+                    return true;
+                }
+
+                error = Interop.Sys.GetLastError();
+                return false;
+            }
+
+            // Try to clone the file immediately, this will only succeed if the
+            // destination doesn't exist, so we don't worry about locking for this one.
+            int flags = Interop.@libc.CLONE_ACL;
+            Interop.Error error;
+            if (TryCloneFile(sourceFullPath, destFullPath, flags, out error))
+            {
+                cloned = true;
+                return;
+            }
+
+            // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+            // This will disable them and allow trying again (a maximum of 1 time).
+            if (error == Interop.Error.EINVAL)
+            {
+                flags = 0;
+                if (TryCloneFile(sourceFullPath, destFullPath, flags, out error))
+                {
+                    cloned = true;
+                    return;
+                }
+            }
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Delete the destination. This should fail on directories. Get a lock to the dest file to ensure we don't copy onto it when
+                // it's locked by something else, and then delete it. It should also fail if destination == source since it's already locked.
+                try
+                {
+                    using SafeFileHandle? dstHandle = SafeFileHandle.Open(destFullPath, FileMode.Open, FileAccess.ReadWrite,
+                        FileShare.None, FileOptions.None, preallocationSize: 0, createOpenException: CreateOpenExceptionForCopyFile);
```

Comment:
I'm pretty sure there's a subtle issue here with copying a file onto a symlink of itself, will open a PR to fix this and add a test for it. This was dealt with in earlier versions of the code, but it seems to have gotten lost along the way.

Reply:
The code uses `FileSafeHandle` just like before which will lock the target of the symlink, just like before this PR.

What makes you think there's an issue now?

Reply:
Upon re-testing it, I think I tested it wrong. It does seem to also not work in net6.0 and net7.0, so will resolve this comment. It does seem a bit odd to lock the target of the symlink though.

Reply:
> It does seem a bit odd to lock the target of the symlink though.

It makes sense because the locking is meant to control access to the file content, which is the target of the link.

Reply:
It doesn't make any sense at all to me. If I have:
'a.txt' a file
'b.txt' a link to 'c.txt'
then I don't see why I would need to lock 'c.txt' to to be able to overwrite it with 'a.txt', it's not part of the operation. What if I didn't open 'c.txt' through 'b.txt' even, now there's unrelated locking going on imo.

Reply:
I didn't even know it did that, and it's not at all what I would expect at all, nor does the documentation even remotely imply any weird behaviour like that.
https://learn.microsoft.com/en-us/dotnet/api/system.io.file.copy?view=net-7.0#system-io-file-copy(system-string-system-string-system-boolean)

Reply:
This is historical. Because symbolic links are not prevalent on Windows, the implementation of most APIs didn't/doesn't consider how it works on symbolic links.

Reply:
Is this something we need to fix? Presumably a quick call to `lstat` (or something cheaper if you know of one) to get `st_mode` would do so we can check the file type.

Reply:
Actually, I think we already get the mode when we open the destination file don't we? We could just use `lstat` instead instead, and then die if it's a symlink.

Reply:
We really need some tests for all this esoteric behaviour if we want to preserve it

Reply:
PR: https://github.com/dotnet/runtime/pull/87600